File: C++/RP2350/src/motor.cpp
#include "motor.hpp"

Motor::Motor(int IN1, int IN2) {
    _pin1 = IN1;
    _pin2 = IN2;
    _isReversed = false;
}

void Motor::setup() {
    pinMode(_pin1, OUTPUT);
    pinMode(_pin2, OUTPUT);
    stop();
}

void Motor::stop() {
    analogWrite(_pin1, 0);
    analogWrite(_pin2, 0);
    digitalWrite(_pin1, LOW);
    digitalWrite(_pin2, LOW);
}

void Motor::reverse() {
    _isReversed = !_isReversed;
}

void Motor::setSpeed(int speed) {
    int val = map(constrain(speed, 0, 100), 0, 100, 0, 255);
    if (!_isReversed) {
        digitalWrite(_pin2, LOW);
        analogWrite(_pin1, val);
    } else {
        digitalWrite(_pin1, LOW);
        analogWrite(_pin2, val);
    }
}

File: src/motor.py
import Jetson.GPIO as GPIO

class Motor:
    # in1 is PWM, in2 is DIR
    def __init__(self, in1: int, in2: int):
        # Setup pins
        try:
            GPIO.setup(in1, GPIO.OUT)
            GPIO.setup(in2, GPIO.OUT)
        except Exception as e:
            print(f"Error occurred: {e}")

        self.IN1 = in1
        self.IN2 = in2
        self.speed = 0
        self.pwm = GPIO.PWM(self.IN1, 1000)
        self.pwm.start(0)

    def set_speed(self, speed: int):
        self.speed = speed
        GPIO.output(self.IN2, GPIO.LOW)
        self.pwm.ChangeDutyCycle(self.speed)

    def reverse(self):
        self.pwm.ChangeDutyCycle(0)
        GPIO.output(self.IN2, GPIO.HIGH)
        self.pwm.ChangeDutyCycle(self.speed)

    def __del__(self):
        self.pwm.stop()